### DISCUSSION TITLE
Updates Rejected because the Tip of Your Current Branch is behind Its Remote Counterpart

### DISCUSSION THREAD

**Discussion by RHH42** [community/community#66050] [Programming Help]
### Body

I'm a raw new user of both GIT and Github.  It's been nearly 50 years - I'm 81 yo - since I was a true programmer (COBOL and BAL).  I'm struggling to understand how Github works.  This is my first post here; so if the topic is inappropriate or in the wrong place, please tell me so I can learn protocol here.

The first time I tried to _push_ my updated GIT _repo_ I got the message referenced in the subject title.  I'm trying to understand what it means and why I got it.  It seems to me that anytime an updated GIT _repo_ is _pushed_ to Github, the code being _pushed_ will be newer than what is in Github.  

Do I first have to merge the Github and GIT _repo_ code somehow?  I can probably figure out how to do that (pull down the old code and manually update it or, preferably, use a GIT function that will do that for me, then _push_ the merged code to Github; but won't this result in the same error?  Must I tell Github that the _push_ is "OK" to replace what's in its _repo_?  How and why do I do that?  My guess is that it will somehow save the older version so that it can be used to restore to a point before changes were applied, which makes a great deal of sense.

Please be patient, I'm coming to my main question (although I need answers to those above too).  If my understanding described above is close to the actual functionality, it seems to me that there must be a more-or-less standard process that should be used periodically as dictates demand.  By that I mean that, for example, before lunch and at end of day a developer would _pull_ his Github code, update it, and _push_it pack to Github indicating somehow to Github that it is the freshest code.

Can someone please respond pointing out the errors in my understanding and where to find answers to my questions?

Thanks for your patience.

### Guidelines

- [X] I have read and understood this category's [guidelines](https://github.com/orgs/community/discussions/51384) before making this post.

**Comment by lab596** [comment]
Hey there! This is what I found on the internet for the error you are encountering:

First a basic recap: 

The error message you're encountering, "Updates Rejected because the Tip of Your Current Branch is behind Its Remote Counterpart," typically means that your local branch is not up-to-date with the remote branch on GitHub.

1. Local vs. Remote: In Git, you have a local repository (on your computer) and a remote repository (on GitHub, in this case). These are separate copies of the same project.

2. Branches: Git allows multiple people to work on the same project simultaneously. Each person usually works on their own branch. A branch is essentially a series of commits (changes) in the project's history.

3. Pushing: When you make changes to your local branch and want to share them with others, you "push" those changes to the remote branch on GitHub.

4. Error: The error you're seeing occurs because someone else has made changes to the remote branch since you last synchronized your local branch with it. Git is designed to prevent you from accidentally overwriting someone else's work.

How to fix the problem: 
1. Pull the Latest Changes: Before pushing your changes, it's essential to ensure your local branch is up-to-date with the remote branch. You can do this by running `git pull origin branch-name`. Replace branch-name with the name of your branch.

2. Resolve Conflicts (If Any): If there are any conflicting changes (changes made to the same lines of code by you and others), Git will ask you to resolve them. You'll need to manually choose which changes to keep.

3. Push Your Changes: After you've pulled the latest changes and resolved any conflicts, you can push your changes to GitHub using `git push origin branch-name`.

4. Pull Request: If you're working collaboratively with others, you'll usually create a "pull request" on GitHub to propose your changes. This allows others to review your code before it's merged into the main branch.

**Reply by RHH42** [reply]
Thanks for the reply.  That is great info.

I discovered since making the post that I did not have my _user.name_ set correctly.  I'm posting this in case it helps someone else.

I went back and changed by my root directory in GIT and did a _git config --globally user.name "XXXXXX"_  then changed the git directory to my working directory.  I then did a _git pull_ read the response then did a git push and all seems to have worked.

**Reply by inmoh7** [reply]
In my case, @RHH42 's solution didn't work. When I tried to do the `git pull` or  `git pull <remote> <branch_name>`  command it threw another error saying 'cannot merge unrelated histories'. I think it was caused by the License file that was created when I created the remote repo. Then I had to force that by, `git pull <remote> <branch_name> --allow-unrelated-histories`. After that pushed the code again and everything worked.
